pkg/updater/eval.go

```diff
@@ -0,0 +1,231 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package updater
+
+import (
+	"regexp"
+	"strings"
+	"sync"
+
+	"github.com/GoogleCloudPlatform/testgrid/metadata/junit"
+	evalpb "github.com/GoogleCloudPlatform/testgrid/pb/custom_evaluator"
+	tspb "github.com/GoogleCloudPlatform/testgrid/pb/test_status"
+)
+
+func strEQ(a, b string) bool { return a == b }
+
+var (
+	res     = map[string]*regexp.Regexp{} // cache regexps for more speed
+	resLock sync.RWMutex
+)
+
+func strReg(val, expr string) bool {
+	resLock.RLock()
+	r, ok := res[expr]
+	resLock.RUnlock()
+	if !ok {
+		var err error
+		if r, err = regexp.Compile(expr); err != nil {
+			r = nil
+		}
+		resLock.Lock()
+		res[expr] = r
+		resLock.Unlock()
+	}
+	if r == nil {
+		return false
+	}
+	return r.MatchString(val)
+}
+
+func strStartsWith(a, b string) bool {
+	return strings.HasPrefix(b, a)
+}
+
+func strContains(a, b string) bool {
+	return strings.Contains(b, a)
+}
+
+func numEQ(a, b float64) bool {
+	return a == b
+}
+
+func numNE(a, b float64) bool {
+	return a != b
+}
+
+// Interface for accessing data about the the result.
+type TestResult interface {
+	// Properties for the the test result.
+	Properties() map[string][]string
+	// Name of the test case
+	Name() string
+	// The number of errors in the test case.
+	Errors() int
+	// The number of failures in the test case.
+	Failures() int
+	// The sequence of exception/error messages in the test case.
+	Exceptions() []string
+}
+
+// CustomStatus evaluates the result according to the rules.
+//
+// Returns nil if no rule matches, otherwise returns the overridden status.
+func CustomStatus(rules []*evalpb.Rule, testResult TestResult) *tspb.TestStatus {
+	for _, rule := range rules {
+		if got := evalProperties(rule, testResult); got != nil {
+			return got
+		}
+	}
+	return nil
+}
+
+type jUnitTestResult struct {
+	Result *junit.Result
+}
+
+func (jr jUnitTestResult) Properties() map[string][]string {
+	if jr.Result == nil || jr.Result.Properties == nil || jr.Result.Properties.PropertyList == nil {
+		return nil
+	}
+	out := make(map[string][]string, len(jr.Result.Properties.PropertyList))
+	for _, p := range jr.Result.Properties.PropertyList {
+		out[p.Name] = append(out[p.Name], p.Value)
+	}
+	return out
+}
+
+func (jr jUnitTestResult) Name() string {
+	return jr.Result.Name
+}
+
+func (jr jUnitTestResult) Errors() int {
+	if jr.Result.Errored != nil {
+		return 1
+	}
+	return 0
+}
+
+func (jr jUnitTestResult) Failures() int {
+	if jr.Result.Failure != nil {
+		return 1
+	}
+	return 0
+}
+
+func (jr jUnitTestResult) Exceptions() []string {
+	options := make([]string, 0, 2)
+	if e := jr.Result.Errored; e != nil {
+		options = append(options, *e)
+	}
+	if f := jr.Result.Failure; f != nil {
+		options = append(options, *f)
+	}
+	return options
+}
+
+func evalProperties(rule *evalpb.Rule, testResult TestResult) *tspb.TestStatus {
```

Comment:
nit: Might be nice to split this into multiple functions.

Reply:
What kind of split are you imagining?

Reply:
Was thinking to pull out a function per type (property, test result field, test result error field), but this isn't that long a function so not actually sure it's better for readability/testability.